### PR TITLE
some prose updates

### PR DESCRIPTION
**Pull request by polinavino**

### CODE REVIEW

File: shelley/chain-and-ledger/formal-spec/crypto-primitives.tex [hunk 1]
```diff
@@ -81,8 +87,20 @@ \section{Cryptographic primitives}
...
+significance to the system. For example, they will be the only ones able to
+propose software and protocol updates. For more on the decentralization process,
+see \ref{sec:new-epoch-trans}.
+
 When we get to the blockchain layer validation, we will use
-key evolving signatures according to the MMM scheme \citep{cryptoeprint:2001:034}.
+key evolving signatures (KES) according to the MMM scheme \citep{cryptoeprint:2001:034}.
+These signature schemes provide forward cryptographic security, meaning that a
+compromised key does make it easier for an adversary to forge a signature that
+allegedly had been signed in the past.
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/crypto-primitives.tex]: a compromised key makes it easier for the adversary ?

File: shelley/chain-and-ledger/formal-spec/intro.tex [hunk 2]
```diff
@@ -31,17 +31,20 @@ \section{Introduction}
 \item[Delegation] Validity of delegation certificates, which delegate
   block-signing rights.
 \item[Stake] Staking rights associated to an address.
+\item[Rewards] Reward calculation and distribution.
+\item[Updates] The update mechanism for Shelley protocol parameters and software
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/intro.tex]:
```suggestion
\item[Updates] The update mechanism for Shelley protocol parameters and software.
```

File: shelley/chain-and-ledger/formal-spec/transactions.tex [hunk 3]
```diff
@@ -6,28 +6,41 @@ \section{Transactions}
 
 \begin{itemize}
   \item A set of transaction inputs.
-    The $\TxIn$ derived type identifies an output from a previous transaction.
+    This derived type identifies an output from a previous transaction.
     It consists of a transaction id and an index to uniquely identify the output.
-  \item An indexed collection of transaction outputs.
+  \item  An indexed collection of transaction outputs.
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/transactions.tex]: please undo this change

File: shelley/chain-and-ledger/formal-spec/transactions.tex [hunk 4]
```diff
@@ -6,28 +6,41 @@ \section{Transactions}
...
   \item A time to live. A transaction will be deemed invalid if processed after this slot.
   \item A mapping of reward account withdrawals.  The type $\Wdrl$ is a finite map that maps
     a reward address to the coin value to be withdrawn. The coin value must be equal
     to the full value contained in the account. Explicitly stating these values ensures
     that error messages can be precise about why a transaction is invalid.
-  \item Update proposals, a pair of mappings from genesis keys to proposals.
-    The first mapping is for protocol parameters and the second one is for application versions.
+    For reward calculation rules, see Section \ref{sec:reward-overview},
+    and for the rule for collecting rewards, see Section \ref{sec:utxo-trans}.
+  \item Update proposals for protocol parameters and software
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/transactions.tex]:
```suggestion
  \item Update proposals for protocol parameters and software.
```

File: shelley/chain-and-ledger/formal-spec/transactions.tex [hunk 5]
```diff
@@ -6,28 +6,41 @@ \section{Transactions}
...
-A transaction, $\Tx$, is a transaction body together with:
+
+The update proposal type $\Update$ is a pair of
+
+\begin{itemize}
+  \item $\PPUpdate$, the protocol parameter upates
+  \item $\AVUpdates$, the updates to Shelley software (applications)
+\end{itemize}
+
+Note that both are finite maps are indexed by the entities proposing the given
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/transactions.tex]:
```suggestion
Note that both of these finite maps are indexed by the entities proposing the given
```

File: shelley/chain-and-ledger/formal-spec/transactions.tex [hunk 6]
```diff
@@ -6,28 +6,41 @@ \section{Transactions}
...
+update, the $\VKeyGen$ keys. This is because updates
+can only be proposed by the owners of the genesis keys.
+(see Section \ref{sec:update}).
+A transaction, $\Tx$, consists of:
 
 \begin{itemize}
+  \item The transaction body
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/transactions.tex]:
```suggestion
  \item The transaction body.
```

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 7]
```diff
@@ -1,26 +1,56 @@
...
+The governance process includes a mechanism for core nodes to propose updates, and
+a voting mechanism all stake holders can participate in. In this chapter we
+outline rules for genesis keys \textit{proposing} both protocol parameter
+and application version updates, as well as voting on whether a particular
+software update is an acceptable future option (again, only genesis keys vote on this).
+For rules regarding the \textit{adoption} of protocol
+parameter updates, see \ref{sec:pparam-update}. For rules regarding
+adoption of new software versions see \ref{sec:software-updates}.
+
+This chapter does not discuss authontication of update proposals.
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/update.tex]:
```suggestion
This chapter does not discuss authentication of update proposals.
```

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 8]
```diff
@@ -1,26 +1,56 @@
...
-the $\mathsf{AVUP}$ transition for the application versions.
-Note that $\PPUpdate$, $\AVUpdate$, and $\Update$ were defined in
-Figure~\ref{fig:defs:utxo-shelley}, and that $\fun{pvCanFollow}$
-was defined in \cite{byron_ledger_spec}.
+The transition type $\mathsf{PPUP}$ is for proposing updates to protocol
+parameters, see Figure \ref{fig:ts-types:pp-update} (for the corresponding rules,
+see Figure \ref{fig:rules:pp-update}). The environment for PP updates contains
+the value $\var{dms}$. We only use here the domain of the $\var{dms}$ finite map
+to access the list of genesis keys,
+but this is the variable in the ledger state that contains these keys.
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/update.tex]: These phrases sound a bit convoluted, maybe merge and simplify them.

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 9]
```diff
@@ -86,33 +116,34 @@ \section{Updates}
...
-\item In the case of \var{slot} being smaller than
-  $\fun{firstSlot}~((\fun{epoch}~\var{slot}) + 1) - \fun{SlotsPrior}$, there is
-  a \em{PPUpdateTooEarly} failure.
-\item In the case of \var{pup} being non-empty, if the check $\dom pup \subseteq
-  \dom dms$ fails, there is a \em{NonGenesisUpdate} failure as only genesis keys
-  can be used in the protocol parameter update.
-\item If a protocol parameter update in \var{pup} cannot follow the current
-  protocol paramter, there is a \em{PVCannotFollow} failure.
-\end{itemize}
+
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/update.tex]: please remove new line here

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 10]
```diff
@@ -86,33 +116,34 @@ \section{Updates}
...
+Recall here that $\type{T}^?$ is similar to the \texttt{Maybe T} type in Haskell,
+and $\Nothing$ represents \texttt{Nothing}.
+
 The function $\fun{validAV}$ uses three functions from \cite{byron_ledger_spec}, namely
-$\fun{apNameValid}$, $\fun{svCanFollow}$, and $\fun{sTagValid}$.
+$\fun{apNameValid}$, $\fun{svCanFollow}$, and $\fun{sTagValid}$. It determines
+whether an application version is valid, given its name, version number,
+metadata, and a set of applications to compare against.
+The function $\fun{newAVs}$ adds the most recent valid application
+versions to (a finite map of) applications (using right override).
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/update.tex]: why the parentheses here?

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 11]
```diff
@@ -154,7 +185,64 @@ \section{Updates}
...
+is considered an update upcoming in the future, and is stored associated
+with a slot which is $\SlotsPrior$ in the future, to be implemented at that time.
+All other update proposals are scrapped.
 
+The AVUP rule has one predicate failure:
+\begin{itemize}
+\item In the case of \var{aup} being non-empty, if the check $\dom aup \subseteq
+  \dom dms$ fails, there is a \em{NonGenesisUpdate} failure as only genesis keys
+  can be used in the application version update.
+\end{itemize}
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/update.tex]: in the above I think there is an `\end{itemize}` missing somewhere

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 12]
```diff
@@ -154,7 +185,65 @@ \section{Updates}
...
+  \item AV-Update-No-Consensus: rule
+is for the case where a voting consensus was not reached on a proposed software
+update after combining the set of new (signal) and existing (state) proposals
+by a right override. In this case, we simply use this combined new value
+of update proposal (votes) as the state UP value.
+  \item AV-Update-Consensus: applies when voting on update proposals by genesis keys
+reaches consensus. In this case, the proposal which was voted on in this slot
+is considered an update upcoming in the future, and is stored associated
+with a slot which is $\SlotsPrior$ in the future, to be implemented at that time.
+All other update proposals are scrapped.
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/update.tex]: you need an `\end{itemize}` here

**Comment by polinavino** [comment]
Oops - the software-update file is not finished, i didnt mean to commit it. I'll commit a new version in an hr or so.

File: shelley/chain-and-ledger/formal-spec/software-updates.tex [hunk 13]
```diff
@@ -1,27 +1,34 @@
...
-    they are using in the blocks they produce.
-  \item The core nodes monitor how many block headers contain the new protocol version.
-    Once enough blocks have the new version, the core nodes post another protocol parameter
-    update to increas the protocol version.
-    If this proposal is accepted, it will be changed at the start of the new epoch and
-    the current software then makes use of new rules.
+  \item New software is ready for downloading.
+  \item The core nodes propose and vote on changing the
+    \textit{application version}.
+    The distinction between protocol and software updates, as well
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/software-updates.tex]: I'm not sure what you mean by the distinction between protocol and software updates. Maybe don't mention protocol parameter updates, but just provide a link to the application version rules?

File: shelley/chain-and-ledger/formal-spec/software-updates.tex [hunk 14]
```diff
@@ -1,27 +1,34 @@
...
+    as the voting and proposal of AV updates, is discussed
+    in Section \ref{sec:update}.
+  \item If accepted, the proposal is stored on the blockchain as the future AV
+    $\var{favs} = s \mapsto ap$,
+    indexed by the slot number $s$ in which $ap$
+    is scheduled to become the official version.
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/software-updates.tex]: note the `ap` is a bunch of applications, and there is an official version of each one.

File: shelley/chain-and-ledger/formal-spec/software-updates.tex [hunk 15]
```diff
@@ -1,27 +1,34 @@
...
+  \item The users can now download the applications listed in $ap$
+  \item In slot $s$, $ap$ becomes the official version, stored as the value
+  $\var{avs}$ on the blockchain
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/software-updates.tex]: after slot $s$, the next block that comes in triggers updating the app versions. technically, it could be that there is a long enough gap that one whole version is skipped, but this is unlikely. anyway, I worry that going too deep into these details makes it confusing, and we can talk about proposals being accepted, etc.

File: shelley/chain-and-ledger/formal-spec/software-updates.tex [hunk 16]
```diff
@@ -1,27 +1,34 @@
...
+  \item All nodes running official Cardano software update their software
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/software-updates.tex]: I'm not sure why we need to talk about "official" software. what's relevant is that nodes can see when the versions change.

**Comment by kevinhammond** [comment]
I think might be supposed to be the difference between major and minor versions?

> On 3 Sep 2019, at 22:55, Jared Corduan <notifications@github.com> wrote:
> 
> @JaredCorduan commented on this pull request.
> 
> In shelley/chain-and-ledger/formal-spec/software-updates.tex:
> 
> > -    will change the application version at the start of the next epoch.
> -  \item All nodes see a new application version and update their software.
> -    Note that the new software should be compatible with the older version, and the new
> -    ledger rules will not be applied as long as the protocol version in the
> -    protocol parameters is not increased.
> -  \item Every block producer includes the protocol version corresponding to the software
> -    they are using in the blocks they produce.
> -  \item The core nodes monitor how many block headers contain the new protocol version.
> -    Once enough blocks have the new version, the core nodes post another protocol parameter
> -    update to increas the protocol version.
> -    If this proposal is accepted, it will be changed at the start of the new epoch and
> -    the current software then makes use of new rules.
> +  \item New software is ready for downloading.
> +  \item The core nodes propose and vote on changing the
> +    \textit{application version}.
> +    The distinction between protocol and software updates, as well
> 
> I'm not sure what you mean by the distinction between protocol and software updates. Maybe don't mention protocol parameter updates, but just provide a link to the application version rules?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 

Best Wishes,
Kevin

--------

E: kevinhammond.net, W: http://www.kevinhammond.net	

This email reflects the opinions of the individual concerned, may contain confidential or copyright information that should not be copied or distributed without permission, may be of a private or personal nature unless explicitly indicated otherwise, and should not under any circumstances be taken as an official statement of policy or procedure.

File: shelley/chain-and-ledger/formal-spec/protocol-parameters.tex [hunk 17]
```diff
@@ -16,7 +16,7 @@ \section{Protocol Parameters}
 representing the number of slots in an epoch/KES period (for a brief explanation
 of a KES period, see Section \ref{sec:crypto-primitives-shelley}). Cycles will be used in
 the key evolving signatures. As global constants, these values can only be
-changed by updating the software.
+changed by updating the software. the
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/protocol-parameters.tex]: ?

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 18]
```diff
@@ -10,9 +10,9 @@ \section{Update Proposal Mechanism}
 software update is an acceptable future option (again, only genesis keys vote on this).
 For rules regarding the \textit{adoption} of protocol
 parameter updates, see \ref{sec:pparam-update}. For rules regarding
-adoption of new software versions see \ref{sec:software-updates}.
+adoption of new software versions see \ref{sec:software-updates}. \newline
 
-\underline{Distinction between protocol parameter and application updates.}
+\noindent \underline{Distinction between protocol parameter and application updates.}
 \begin{itemize}
 \item \textbf{Protocol parameters} (excluding $\ProtVer$, explained below): constants
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/update.tex]: why distinguish the protocol version? it's just a protocol parameter like the others.

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 19]
```diff
@@ -82,9 +85,7 @@ \section{Update Proposal Mechanism}
 for a transaction, see \ref{sec:witnesses-shelley}.
 
 \begin{note}
-\underline{Genesis Key Delegations}
-\newline
-\noindent The environment for both PP and AV updates contains
+\textbf{Genesis Key Delegations.} The environment for both PP and AV updates contains
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/update.tex]: The core nodes all currently delegated, and cannot "undelegate".

File: shelley/chain-and-ledger/formal-spec/crypto-primitives.tex [hunk 20]
```diff
@@ -85,10 +85,26 @@ \section{Cryptographic primitives}
...
 
+In KES, the public verification key stays constant, but the
+corresponding private key evolves incrementally. For this reason, KES
+verification keys are indexed by integers representing the step in the key's
+evolution. This evolution step parameter is also an additional parameter needed
+for the signing (denoted by $\signEv$) and verification
+(denoted by $\verifyEv$) functions.
+
+Since the private key evolves incrementally in a KES scheme, the ledger rules
+must dictate the amount of time between incrementations. This is referred to as
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/crypto-primitives.tex]: another way to say this is that the ledger is requiring the pool operators to evolve their keys once per `KESPeriod`-many slots.

File: shelley/chain-and-ledger/formal-spec/notation.tex [hunk 21]
```diff
@@ -40,9 +40,21 @@ \section{Notation}
   \item[Option type] An option type in type $A$ is denoted as $A^? = A + \Nothing$. The
     $A$ case corresponds to the case when there is a value of type $A$ and the $1$
     case corresponds to the case when there is no value.
-
 \end{description}
 
+\begin{note}
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/notation.tex]: maybe we can just list the let binding notation in the same itemized list with the other notation instead of making a note?

File: shelley/chain-and-ledger/formal-spec/software-updates.tex [hunk 22]
```diff
@@ -1,22 +1,34 @@
...
+  automatically, as soon as they see the updated official version on the ledger.
+  \item This new software should be implementing the ledger rules in the
+  PV currently in the PP's on the ledger
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/software-updates.tex]: Is this a rewording of what used to be
```Note that the new software should be compatible with the older version, and the new ledger rules will not be applied as long as the protocol version in the protocol parameters is not increased.
```?

File: shelley/chain-and-ledger/formal-spec/transactions.tex [hunk 23]
```diff
@@ -19,15 +19,52 @@ \section{Transactions}
...
+(public verification) keys, and these are a distinct type than the usual
+user verification keys, $\VKey$. Genesis keys are the keys belonging to the federated
+nodes running the Cardano system currently (also referred to as core nodes).
+
+Currently, updates
+can only be proposed and voted on by the owners of the genesis keys.
+The process of decentralization will result in the core nodes gradually giving up
+some of their power and responsibilities to the other system nodes.
+It may be the case in the future that the update proposal mechanism will
+be decentralized as well, and allow all nodes to participate.
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/transactions.tex]: The core nodes _will_ eventually give up _all_ of their elevated privileges, it's just a matter of when (but not Shelley).

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 24]
```diff
@@ -1,26 +1,140 @@
...
+\begin{itemize}
+\item \textbf{Protocol parameters} (excluding $\ProtVer$, explained below): constants
+currently used in ledger calculations according to the rules described in this document
+\begin{itemize}
+\item[$\circ$] Stored on the leger
+\item[$\circ$] Changed only at epoch boundary
+\item[$\circ$] Changed according to the genesis key votes, which are also stored on the ledger
+\item[$\circ$] All nodes automatically adopt new values (this mechanism is an explicit
+ part of the ledger rules, and is included in the epoch boundary transition)
+\item[$\circ$] Has no effect on ledger transition rules
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/update.tex]: If we change, say, the max transaction size, then in a sense there is an effect on the transition rules. I guess you mean some kind of logic change?

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 25]
```diff
@@ -1,26 +1,140 @@
...
+\item[$\circ$] Switching to using these new rules is mandatory in the sense that
+if the nodes do not have the applications implementing them, this
+may prevent a user from reading and
+writing to the blockchain, e.g changing the
+transaction data structure would cause such a problem
+\end{itemize}
+\item \textbf{Applications} The version of the software the nodes run,
+as well as the related system tags and metadata
+\begin{itemize}
+\item[$\circ$] These updates cannot be forced
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/formal-spec/update.tex]: we cannot force the users to actually upgrade their software, but the data on the chain for the application versions is in a sense forced.

File: shelley/chain-and-ledger/formal-spec/crypto-primitives.tex [hunk 26]
```diff
@@ -158,14 +157,15 @@ \section{Cryptographic primitives}
...
+a valid combination for the script, otherwise it returns $\mathsf{False}$.
 
-Use cases for this are for example:
+Use cases for this are, for example, if we want to impose any of the following
+requirements on spending an output:
 \begin{itemize}
-\item Require two parties to sign.
-\item Require either of two parties to sign.
-\item Generally, require $m$ out of $n$ ($m \leq n$) parties to sign.
+\item two parties to required sign;
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/crypto-primitives.tex]: are required here?

File: shelley/chain-and-ledger/formal-spec/notation.tex [hunk 27]
```diff
@@ -49,11 +49,11 @@ \section{Notation}
 
   Then, in subsequent expressions, it is more convenient to write simply $s$.
   It is not meant to shadow variables, and if it does, there is likely a problem with the
-  rules that must be addressed.
+  rules that must be addressed.
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/notation.tex]: please remove the added space

File: shelley/chain-and-ledger/formal-spec/software-updates.tex [hunk 28]
```diff
@@ -1,22 +1,101 @@
...
+which applications will automatically follow
+\end{itemize}
+
+Applications must sometimes support \textit{several different versions}
+of ledger rules in order to accommodate the timely switch of the $\ProtVer$ at the
+epoch boundary. Usually, the currently accepted protocol version, and next the
+version they are ready to upgrade to (that their application versions can
+implement).
+The newest protocol version a node is ready to use is included in the block
+header of the blocks it makes, see \ref{sec:defs-blocks}. This is either
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/software-updates.tex]: make "produces" instead of "makes" ?

File: shelley/chain-and-ledger/formal-spec/software-updates.tex [hunk 29]
```diff
@@ -1,22 +1,101 @@
...
+  see Section \ref{sec:ledger-trans}.
+  \item All nodes see a new application version on the ledger and update their
+  software.
 \end{enumerate}
+
+Note that if there is a \textit{new protocol version} implemented by this new
+software, the core nodes monitor how many nodes are ready to use the new
+protocol version via the block headers.
+Once enough nodes are ready for the new protocol version, this
+may now be updated as well (by the mechanism in Section \ref{sec:update}).
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/software-updates.tex]: ... mechanism described in ...

File: shelley/chain-and-ledger/formal-spec/crypto-primitives.tex [hunk 30]
```diff
@@ -102,8 +102,9 @@ \section{Cryptographic primitives}
 (denoted by $\verifyEv$) functions.
 
 Since the private key evolves incrementally in a KES scheme, the ledger rules
-must dictate the amount of time between incrementations. This is referred to as
-the KES period, and is the constant $\KESPeriod$. 
+require the pool operators to evolve their keys once per KES period-slots.  
+This is the constant $\KESPeriod$.
+
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/crypto-primitives.tex]: unneeded extra line

File: shelley/chain-and-ledger/formal-spec/update.tex [hunk 31]
```diff
@@ -87,12 +87,12 @@ \section{Update Proposal Mechanism}
...
-This is the genesis key delegations. During the Byron era, they are expected
-to all have delegated to some $\VKey$, and these delegations are inherited
+This is the genesis key delegations. During the Byron era, they are all
+already delegated to some $\VKey$, and these delegations are inherited
 through the Byron-Shelley transition (see \ref{sec:byron-to-shelley}).
 These delegations can be updated as described in \ref{sec:delegation-shelley},
-but those genesis keys that do not delegate to anything are not able to vote,
-since the rules below only count the votes of delegated keys.
+but there is no mechanism for them to un-delegate or for the keys they delegate
+to to retire (unlike regular stake pools).
```
Comment (mgudemann) [inline shelley/chain-and-ledger/formal-spec/update.tex]: replicated "to"
